presto-spi/src/main/java/com/facebook/presto/spi/memory/Caches.java

```diff
@@ -13,29 +13,86 @@
  */
 package com.facebook.presto.spi.memory;
 
+import java.util.Arrays;
+
 public class Caches
 {
-    private static ByteArrayPool byteArrayPool;
+    private static ArrayPool<byte[]> byteArrayPool;
     private static ByteArrayPoolCacheAdapter byteArrayPoolCacheAdapter;
+    private static class BooleanArrayAllocator
+            extends ArrayPool.Allocator<boolean[]>
+    {
+        @Override
+        boolean[] alloc(int size)
+        {
+            return new boolean[size];
+        }
+
+        @Override
+        void init(boolean[] array)
+        {
+            Arrays.fill(array, false);
+        }
+
+        @Override
+        int getLength(boolean[] array)
+        {
+            return array.length;
+        }
+    }
+
+    private static class ByteArrayAllocator
+            extends ArrayPool.Allocator<byte[]>
+    {
+        @Override
+        byte[] alloc(int size)
+        {
+            return new byte[size];
+        }
+
+        @Override
+        void init(byte[] array)
+        {
+            Arrays.fill(array, (byte) 0);
+        }
+
+        @Override
+        int getLength(byte[] array)
+        {
+            return array.length;
+        }
+    }
+
+    private static ArrayPool<boolean[]> booleanArrayPool;
+
+    public static ArrayPool<boolean[]> getBooleanArrayPool()
```

Comment:
Why not initialize all these at startup? Is lazy initialization and extra locking needed?